File: create_png_files.py
#!/usr/bin/env python3
"""
Create valid solid-color PNG files to replace corrupted ones.
This script uses only Python standard library (struct and zlib).
"""
import os
import struct
import zlib


def create_solid_color_png(filepath, rgb_tuple, size=(192, 192)):
    """
    Create a solid-color PNG file using raw PNG format.
    
    Args:
        filepath: Path where to save the PNG
        rgb_tuple: (R, G, B) color tuple (0-255)
        size: (width, height) in pixels
    """
    r, g, b = rgb_tuple
    width, height = size
    
    # PNG signature
    png_sig = b'\x89PNG\r\n\x1a\n'
    
    # IHDR chunk (image header)
    ihdr_data = struct.pack('>IIBBBBB', width, height, 8, 2, 0, 0, 0)  # 8-bit RGB
    ihdr_crc = zlib.crc32(b'IHDR' + ihdr_data) & 0xffffffff
    ihdr_chunk = struct.pack('>I', 13) + b'IHDR' + ihdr_data + struct.pack('>I', ihdr_crc)
    
    # IDAT chunk (image data)
    raw_data = b''
    for y in range(height):
        raw_data += b'\x00'  # filter type for each scanline
        for x in range(width):
            raw_data += bytes([r, g, b])
    
    compressed_data = zlib.compress(raw_data)
    idat_crc = zlib.crc32(b'IDAT' + compressed_data) & 0xffffffff
    idat_chunk = struct.pack('>I', len(compressed_data)) + b'IDAT' + compressed_data + struct.pack('>I', idat_crc)
    
    # IEND chunk (end)
    iend_crc = zlib.crc32(b'IEND') & 0xffffffff
    iend_chunk = struct.pack('>I', 0) + b'IEND' + struct.pack('>I', iend_crc)
    
    # Write PNG file
    os.makedirs(os.path.dirname(filepath), exist_ok=True)
    with open(filepath, 'wb') as f:
        f.write(png_sig + ihdr_chunk + idat_chunk + iend_chunk)
    
    return os.path.getsize(filepath)


def main():
    """Create the three placeholder PNG files."""
    base_path = 'app/src/main/res/mipmap-hdpi'
    
    files_to_create = [
        ('color_q7.png', (255, 107, 107)),    # Red
        ('color_q12.png', (78, 205, 196)),    # Teal
        ('color_q13.png', (255, 230, 109)),   # Yellow
    ]
    
    print("Creating placeholder PNG files...")
    for filename, color_rgb in files_to_create:
        filepath = os.path.join(base_path, filename)
        size_bytes = create_solid_color_png(filepath, color_rgb)
        print(f"✓ Created {filename} ({size_bytes} bytes, RGB{color_rgb})")
    
    print("\nAll PNG files created successfully!")


if __name__ == '__main__':
    main()

File: create_png_files.js
#!/usr/bin/env node
/**
 * Create valid solid-color PNG files to replace corrupted ones.
 * Uses Node.js built-in modules only.
 */
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

function createSolidColorPNG(filepath, rgbTuple, size = [192, 192]) {
  const [r, g, b] = rgbTuple;
  const [width, height] = size;
  
  // PNG signature
  const pngSig = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
  
  // IHDR chunk (image header)
  const ihdrData = Buffer.alloc(13);
  ihdrData.writeUInt32BE(width, 0);
  ihdrData.writeUInt32BE(height, 4);
  ihdrData[8] = 8;   // bit depth
  ihdrData[9] = 2;   // color type (RGB)
  ihdrData[10] = 0;  // compression method
  ihdrData[11] = 0;  // filter method
  ihdrData[12] = 0;  // interlace method
  
  const ihdrCrc = crc32(Buffer.concat([Buffer.from('IHDR'), ihdrData]));
  const ihdrChunk = Buffer.concat([
    uint32BE(13),
    Buffer.from('IHDR'),
    ihdrData,
    uint32BE(ihdrCrc)
  ]);
  
  // IDAT chunk (image data)
  let rawData = Buffer.alloc(0);
  for (let y = 0; y < height; y++) {
    const scanline = Buffer.alloc(width * 3 + 1);
    scanline[0] = 0;  // filter type
    for (let x = 0; x < width; x++) {
      scanline[1 + x * 3] = r;
      scanline[1 + x * 3 + 1] = g;
      scanline[1 + x * 3 + 2] = b;
    }
    rawData = Buffer.concat([rawData, scanline]);
  }
  
  const compressedData = zlib.deflateSync(rawData);
  const idatCrc = crc32(Buffer.concat([Buffer.from('IDAT'), compressedData]));
  const idatChunk = Buffer.concat([
    uint32BE(compressedData.length),
    Buffer.from('IDAT'),
    compressedData,
    uint32BE(idatCrc)
  ]);
  
  // IEND chunk (end)
  const iendCrc = crc32(Buffer.from('IEND'));
  const iendChunk = Buffer.concat([
    uint32BE(0),
    Buffer.from('IEND'),
    uint32BE(iendCrc)
  ]);
  
  // Write PNG file
  const dir = path.dirname(filepath);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  
  fs.writeFileSync(filepath, Buffer.concat([pngSig, ihdrChunk, idatChunk, iendChunk]));
  return fs.statSync(filepath).size;
}

function uint32BE(value) {
  const buf = Buffer.alloc(4);
  buf.writeUInt32BE(value, 0);
  return buf;
}

function crc32(buf) {
  let crc = 0xffffffff;
  for (let i = 0; i < buf.length; i++) {
    crc = crc ^ buf[i];
    for (let j = 0; j < 8; j++) {
      crc = (crc >>> 1) ^ ((crc & 1) ? 0xedb88320 : 0);
    }
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function main() {
  const basePath = 'app/src/main/res/mipmap-hdpi';
  
  const filesToCreate = [
    ['color_q7.png', [255, 107, 107]],    // Red
    ['color_q12.png', [78, 205, 196]],    // Teal
    ['color_q13.png', [255, 230, 109]],   // Yellow
  ];
  
  console.log('Creating placeholder PNG files...');
  for (const [filename, colorRgb] of filesToCreate) {
    const filepath = path.join(basePath, filename);
    const sizeBytes = createSolidColorPNG(filepath, colorRgb);
    console.log(`✓ Created ${filename} (${sizeBytes} bytes, RGB${colorRgb})`);
  }
  
  console.log('\nAll PNG files created successfully!');
}

main();
